fabrics.c

```diff
@@ -505,63 +508,77 @@ int nvmf_discover(const char *desc, int argc, char **argv, bool connect)
 		goto out_free;
 	}
 
-	if (device) {
-		c = nvme_scan_ctrl(r, device);
-		if (c) {
-			/* Check if device matches command-line options */
-			if (!nvme_ctrl_is_discovery_ctrl(c) ||
-			    strcmp(nvme_ctrl_get_transport(c), transport) ||
-			    strcasecmp(nvme_ctrl_get_traddr(c), traddr) ||
-			    (cfg.host_traddr && nvme_ctrl_get_host_traddr(c) &&
-			     strcmp(nvme_ctrl_get_host_traddr(c),
-				    cfg.host_traddr)) ||
-			    (cfg.host_iface && nvme_ctrl_get_host_iface(c) &&
-			     strcmp(nvme_ctrl_get_host_iface(c),
-				    cfg.host_iface)) ||
-			    (trsvcid && nvme_ctrl_get_trsvcid(c) &&
-			     strcmp(nvme_ctrl_get_trsvcid(c), trsvcid))) {
-				fprintf(stderr,
-					"ignoring ctrl device %s, "
-					"command-line options do not match\n",
-					device);
-				c = NULL;
-				persistent = false;
-			} else {
-				/*
-				 * If the controller device is found it must
-				 * be persistent, and shouldn't be disconnected
-				 * on exit.
-				 */
-				persistent = true;
-			}
-		} else {
-			/*
-			 * No controller found, fall back to create one.
-			 * But that controller cannot be persistent.
-			 */
-			fprintf(stderr,
-				"ctrl device %s not found%s\n", device,
-				persistent ? ", ignoring --persistent" : "");
-			persistent = false;
-		}
+	/*
+	 * If no device has been provided or the user insists in a new
+	 * one create a new ctrl.
+	 */
+	if (!device || (persistent && force))
+		goto create_ctrl;
+
+	/* Use provided device as ctrl */
+	c = nvme_scan_ctrl(r, device);
+	if (!c) {
+		ret = errno;
+		goto out_free;
+	}
+
+	/* Check if device matches command-line options */
```

Comment:
@hreinecke - I had a discussion with @igaw regarding the `--device` option and we wanted your opinion. As you know, the `--device` option allows reusing a persistent connection. However, when using this option one must also provide all the other options that would have been used to create the connection (-t, -a, -s, etc). And I believe it's because of this code right here that tries to match all the command-line options.

When using `--device`, wouldn't it make sense to just allow it w/o requiring that all other parameters be specified as well? Here's an example: 
```
nvme discover --device nvme0
```
Currently, this is what one has to do this to reuse a persistent device:
```
nvme discover -t tcp -a 212.10.1.199 -s 8009 -n nqn.1988-11.com.dell:powerstore:00:829bb3af984dBD581792 --device nvme0
```
Any thoughts?

Reply:
Oh well. Over-specification _again_.
I really hate this kind of making-absolutely-sure-and-add-more-failure-cases style of programming.
If we have enough information to uniquely identify something  by all means _use it_ and don't add more parameters.

This goes especially for the persistent discovery controller case; here we have a) the parameters (which in itself are enough to identify the device) and b) the device node (which again is enough to identify the .. well .. device).
So why do I have to specify _both_?
Where's the point in that, other than adding more possible failure cases?

And then there's the '-p' option, which serves to confuse the matters even more, as it really only signals 'create a persistent discovery controller', but remains silent if there already _is_ a persistent discovery controller.

Reply:
Anyway. What we have been agreed upon is to keep the '-p' option to indicate 'Create a persistent discovery controller'.
And if a persistent discovery controller is present '-p' will be a no-op.
(That alone should be a patch, and if only to update the manpage reflecting this scenario.)

Reply:
Additionally there will be a new option '--force' to force the creation of a persistent discovery controller, even if one exists.
(Which is what this patch is about.)

Reply:
And the final question is what you outlined above:
Is 'nvme discover -d <devname>' a valid command?

In my opinion, yes, that should be sufficient. I completely fail to see what we would gain if we require the user to specify all connect options, other than introducing yet more possible failure cases.

Reply:
okay let me try to get the 'nvme discover -d' case working

Reply:
What about just adding a warning and and just use the '-d' discovery controller? Basically we do this for the --persistent  option already. We just ignore it when it doesn't make sense.